### ISSUE TITLE
`sync`: pick up a "provides" package from the command line

### ISSUE THREAD

**Issue by CyberShadow** [AladW/aurutils#837]
Consider: <code>aur sync [bitlbee-libpurple-git](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=bitlbee-libpurple-git) [bitlbee-discord-git](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=bitlbee-discord-git)</code>

`bitlbee-libpurple-git` provides `bitlbee`, and `bitlbee-discord-git` depends on `bitlbee`, so one would expect that `bitlbee-discord-git` is built against `bitlbee-libpurple-git`. However, aurutils pulls in the [`bitlbee`](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=bitlbee) package instead.

Not sure if this is a defect or maybe there is a good reason for this. However, it is at least inconsistent with the behavior when `bitlbee-discord-git` is available in a repository; then, aurutils does not pull in `bitlbee`, but just uses `bitlbee-discord-git` which provides it.

**Comment by AladW** [comment]
https://gitlab.archlinux.org/archlinux/aurweb/-/issues/307

**Comment by AladW** [comment]
With RPCv6, this should be as simple as:
```
{ aur query -t search -b provides - <argv | jq -r '.results[].Name'
  ...
} > filter
```

**Comment by AladW** [comment]
This should work (at the cost of a large file download):
```
$ aur pkglist --info | jq -r '.[] | select(.Provides[]? | contains("libcubeb.so")) | .Name'
```

**Comment by VorpalBlade** [comment]
This is also an issue with .so provides on the command line. Example command which triggered issues:

```
aur sync -c qtforkawesome qtutilities c++utilities syncthingtray
```

qtforkawesome fails to build because (according to alad on IRC):
> &lt;alad&gt; Vorpal, oh, I see now for 1): error: target not found: libqtforkawesome.so - this would likely by solved by using aur-graph to resolve the dependencies a second time, using the .SRCINFO

**Comment by AladW** [comment]
`provides` on the command line are addressed with 9680df6ae9318c5dcbf52a7ee3ac3690521944e5

**Comment by AladW** [comment]
So, with `aur-format` merged you can create a `shlibs` file as follows:
```bash
aur pkglist --info | aur format -f '%n\t%P\n' --delim=, \
    | awk -F'\t' '{ split($2, provs, ","); for (p in provs) { sub(/[<>=].*/, "", provs[p]); printf("%s\t%s\n", provs[p], $1); }; }' \
    | sort -k1,1b -u > shlibs
```
This file can then be used to resolve `provides` to AUR targets. Sample output:
```
libaacs.so      libaacs-git
libacl.so       acl-git
libad9361       libad9361-git
libad9361-git   libad9361-git
libad9361-iio   libad9361-iio-git
libad9361.so    libad9361-git
libadalang      libadalang-git
libadlmidi      libadlmidi-git
libadvo libadvo
libadwaita      libadwaita-git
libadwaita-1.so libadwaita-git
```
It might make sense to remove targets provided by repo packages from this output:
```bash
datamash -W check < shlibs  # check for malformed packages
cut -f1 shlibs | aur repo-filter --sync > shlibs_repo
grep -wvxf shlibs_repo shlibs > shlibs_aur
```
`repo-filter` is however much slower than `format` :thinking: 
```
Benchmark 1: cut -f1 shlibs_all | aur repo-filter --sync
  Time (mean ± σ):     17.253 s ±  0.659 s    [User: 17.452 s, System: 0.689 s]
  Range (min … max):   16.292 s … 18.378 s    10 runs
```

**Comment by VorpalBlade** [comment]
> repo-filter is however much slower than format thinking

Should it not be possible to cache this result for, say, 24 hours? In which case it doesn't really matter. You could set it up as a systemd timer (user one perhaps).

**Comment by AladW** [comment]
Sure but there should be faster alternatives available. For example:
1. Filter out any `provides` which already have an AUR `pkgname` with `aur-pkglist`.
2. Use `expac -S '%P'` and filter out repository `provides` by hand.

**Comment by AladW** [comment]
It took a rewrite in `perl` to solve this fully: https://github.com/AladW/aurutils/pull/1041

Any `provides` specified on the command-line are used transparently. `--table` also adds a `Self` line for any target mentioned on the command-line. Demonstration:
```
# Case 1: primary target specified
$ aur depends bitlbee-discord-git --table
bitlbee-discord-git	bitlbee-discord-git	bitlbee-discord-git	0.4.3-1	Self
bitlbee	bitlbee-discord-git	bitlbee	3.6-2	Depends
```
```
# Case 2: primary target and dependency specified
$ aur depends bitlbee-discord-git bitlbee --table
bitlbee	bitlbee-discord-git	bitlbee	3.6-2	Depends
bitlbee	bitlbee	bitlbee	3.6-2	Self
bitlbee-discord-git	bitlbee-discord-git	bitlbee-discord-git	0.4.3-1	Self
```
```
# Case 3: primary target and provider specified
$ aur depends bitlbee-libpurple-git bitlbee-discord-git --table
bitlbee-discord-git	bitlbee-discord-git	bitlbee-discord-git	0.4.3-1	Self
bitlbee-libpurple-git	bitlbee-discord-git	bitlbee-libpurple-git	3.5.1.r80.gc3599e03-1	Depends
bitlbee-libpurple-git	bitlbee-libpurple-git	bitlbee-libpurple-git	3.5.1.r80.gc3599e03-1	Self
```
Additional logic makes it works with providers that don't have an existing `pkgname`. Demonstration:
```
$ diff -u <(aur depends --table yuzu | sort -u) <(aur depends --table yuzu cubeb-git | sort -u)
--- /dev/fd/63	2022-11-28 15:43:32.151803449 +0100
+++ /dev/fd/62	2022-11-28 15:43:32.151803449 +0100
@@ -1,7 +1,12 @@
+celt	jack	celt	0.11.3-5	MakeDepends
 cpp-httplib-compiled	yuzu	cpp-httplib-compiled	0.11.2-1	Depends
 cpp-httplib-compiled	yuzu	cpp-httplib-compiled	0.11.2-1	MakeDepends
+cubeb-git	cubeb-git	cubeb-git	0.2.r1395.g28c8aa4-1	Self
+cubeb-git	yuzu	cubeb-git	0.2.r1395.g28c8aa4-1	Depends
+cubeb-git	yuzu	cubeb-git	0.2.r1395.g28c8aa4-1	MakeDepends
 dynarmic	yuzu	dynarmic	6.3.1-1	Depends
 dynarmic	yuzu	dynarmic	6.3.1-1	MakeDepends
+jack	cubeb-git	jack	0.126.0-5	MakeDepends
 xbyak	dynarmic	xbyak	6.66-1	MakeDepends
 xbyak	yuzu	xbyak	6.66-1	MakeDepends
 yuzu	yuzu	yuzu	mainline.0.1245-1	Self
```
Note that #1041 does not update `aur-sync` to account for `aur-depends` changes yet. This should be a routine change though.

**Comment by CyberShadow** [comment]
Sweet, nice work!

**Comment by AladW** [comment]
> Note that https://github.com/AladW/aurutils/pull/1041 does not update aur-sync to account for aur-depends changes yet. This should be a routine change though.

This was also implemented in #1041
